test.gr

```diff
@@ -0,0 +1,46 @@
+import Vec
+
+data Bool where True; False
+
+foo : Int -> Int |0..1| -> Int
+foo x |y| = if x > 5 then 0 else y
```

Comment:
I wonder if it would be possible at one stage to mark the postcondition of this function as being `|0..1|` - something like:

```
foo : Int -> Int |0..1| -> Int |0..1|
foo x |y| = if x > 5 then 0 else y
```

I don't fully understand the coeffect stuff yet though, so I might be off the mark here!

Reply:
Hey, thanks for your question!

The thing you wrote almost type checks, however, you need to promote the return value:

```idris
foo : Int -> Int |0..1| -> Int |0..1|
foo x |y| = if x > 5
  then |0| else |y| -- these guys are promoted
```

It's not clear to me yet when you would want to do this. I'll think about it.

Reply:
Oh interesting! I always thought coeffects had to be attached to parameters, and effects to return types, so this surprises me!

> It's not clear to me yet when you would want to do this.

One example:

```
foo 42 (foo 43 1)
```

Without the constraint on the return type, would this type check?

Reply:
> Without the constraint on the return type, would this type check?

Not quite, since the second argument to `foo` is not well-typed. You can fire up granule interactive to check:

```
Granule> :l test.gr
test.gr, interpreted

Granule> foo 43 1
Type error: :1:8:
  Int is not equal to ((Int) |0..1|)

Granule> foo 43 |1|
0
```
The problem is that the second argument to `foo` needs to be boxed. Thus, for your example, you'd want `foo 42 |(foo 43 |1|)|`.

> [...] Quantitative Type Theory and Context Constrained Computation, where one can only ever mark coeffects on parameters.

I think in a language based on QTT you _can_ have coeffects on return values by wrapping them up in a constructor, e.g. (informally)

```
data Box a where
  MkBox : a -omega-> Box a
```

Reply:
> The problem is that the second argument to `foo` needs to be boxed. Thus, for your example, you'd want `foo 42 |(foo 43 |1|)|`.

Ah right! Sorry, still getting my head around the boxing stuff. Boxing aside, the reason I was mentioning it was that if the return type of `foo` was not constrained to `|0..1|`, would that not be a type error if you were to pass that to the second argument of another call to `foo`?

Reply:
I'm not sure what you mean. Can you give me the expression you had in mind?

Reply:
This type checks fine. You are promoting/"boxing" `test1` in the call to `foo` in `test2`. When promoting, you have to prove that you can produce 0..1 copies of `test1`; this is trivial because `test1` is a closed term. If `test1` was binding any variables, then we would have to ensure that those are available the correct number of times.

I hope this makes some sense!

Reply:
Ohhhh! This is working on multiplicities, not on the numeric range 🤦‍♂️